File: packages/python/elizaos/features/advanced_capabilities/actions/schedule_follow_up.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import SCHEDULE_FOLLOW_UP_TEMPLATE
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)
from elizaos.utils.spec_examples import convert_spec_examples  # noqa: F401
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

# Get text content from centralized specs
_spec = require_action_spec("SCHEDULE_FOLLOW_UP")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class ScheduleFollowUpAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        relationships_service = runtime.get_service("relationships")
        follow_up_service = runtime.get_service("follow_up")
        return relationships_service is not None and follow_up_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.features.advanced_capabilities.services.follow_up import FollowUpService
        from elizaos.features.advanced_capabilities.services.relationships import (
            RelationshipsService,
        )

        relationships_service = runtime.get_service("relationships")
        follow_up_service = runtime.get_service("follow_up")

        if not relationships_service or not isinstance(relationships_service, RelationshipsService):
            return ActionResult(
                text="Relationships service not available",
                success=False,
                values={"error": True},
                data={"error": "RelationshipsService not available"},
            )

        if not follow_up_service or not isinstance(follow_up_service, FollowUpService):
            return ActionResult(
                text="Follow-up service not available",
                success=False,
                values={"error": True},
                data={"error": "FollowUpService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ENTITIES"])
        state.values["currentDateTime"] = datetime.utcnow().isoformat()

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=SCHEDULE_FOLLOW_UP_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        if not parsed or not parsed.get("contactName"):
            return ActionResult(
                text="Could not extract follow-up information",
                success=False,
                values={"error": True},
                data={"error": "Failed to parse follow-up info"},
            )

        contact_name = str(parsed.get("contactName", ""))
        scheduled_at_str = str(parsed.get("scheduledAt", ""))
        reason = str(parsed.get("reason", "Follow-up"))
        priority = str(parsed.get("priority", "medium"))
        follow_up_message = str(parsed.get("message", ""))

        # Normalize priority to valid values
        if priority not in ("low", "medium", "high"):
            priority = "medium"

        # Validate scheduled_at
        try:
            scheduled_at = datetime.fromisoformat(scheduled_at_str.replace("Z", "+00:00"))
        except ValueError:
            return ActionResult(
                text="Invalid date format for scheduled time",
                success=False,
                values={"error": True},
                data={"error": "Invalid scheduledAt date format"},
            )

        # Resolve contact via relationships
        contacts = await relationships_service.search_contacts(search_term=contact_name)
        if not contacts:
            return ActionResult(
                text=f"Could not find contact '{contact_name}' in relationships",
                success=False,
                values={"error": True},
                data={"error": f"Contact '{contact_name}' not found"},
            )

        entity_id = contacts[0].entity_id
        await follow_up_service.schedule_follow_up(
            entity_id=entity_id,
            scheduled_at=scheduled_at,
            reason=reason,
            priority=priority,
            message=follow_up_message,
        )

        response_text = f"I've scheduled a follow-up with {contact_name} for {scheduled_at.strftime('%B %d, %Y')}. Reason: {reason}"

        if callback:
            await callback(Content(text=response_text, actions=["SCHEDULE_FOLLOW_UP"]))

        return ActionResult(
            text=response_text,
            success=True,
            values={
                "contactId": str(entity_id),
                "scheduledAt": scheduled_at.isoformat(),
            },
            data={
                "contactId": str(entity_id),
                "contactName": contact_name,
                "scheduledAt": scheduled_at.isoformat(),
                "reason": reason,
                "priority": priority,
            },
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


schedule_follow_up_action = Action(
    name=ScheduleFollowUpAction.name,
    similes=ScheduleFollowUpAction().similes,
    description=ScheduleFollowUpAction.description,
    validate=ScheduleFollowUpAction().validate,
    handler=ScheduleFollowUpAction().handler,
    examples=ScheduleFollowUpAction().examples,
)

File: packages/typescript/src/features/advanced-capabilities/actions/scheduleFollowUp.ts
import { findEntityByName } from "../../../entities.ts";
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { scheduleFollowUpTemplate } from "../../../prompts.ts";
import type { FollowUpService } from "../../../services/followUp.ts";
import type { RelationshipsService } from "../../../services/relationships.ts";
import {
	extractScheduleFollowUpResponseFromText,
	type ParsedScheduleFollowUpResponse,
} from "../../shared/schedule-follow-up-response.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { asUUID, ModelType } from "../../../types/index.ts";
import {
	composePromptFromState,
	parseJSONObjectFromText,
	parseKeyValueXml,
} from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("SCHEDULE_FOLLOW_UP");
const FOLLOW_UP_KEYWORDS = getValidationKeywordTerms(
	"action.scheduleFollowUp.request",
	{
		includeAllLocales: true,
	},
);

export const scheduleFollowUpAction: Action = {
	name: spec.name,
	description: spec.description,
	similes: spec.similes ? [...spec.similes] : [],
	examples: (spec.examples ?? []) as ActionExample[][],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> => {
		if (
			runtime.actions.some(
				(action) =>
					action.name === "OWNER_RELATIONSHIP" ||
					action.name === "RELATIONSHIP",
			)
		) {
			return false;
		}

		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;
		const followUpService = runtime.getService("follow_up") as FollowUpService;

		if (!relationshipsService || !followUpService) {
			logger.warn("[ScheduleFollowUp] Required services not available");
			return false;
		}

		const messageText = message.content.text ?? "";
		if (!messageText) return false;
		return findKeywordTermMatch(messageText, FOLLOW_UP_KEYWORDS) !== undefined;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult | undefined> => {
		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;
		const followUpService = runtime.getService("follow_up") as FollowUpService;

		if (!relationshipsService || !followUpService) {
			throw new Error("Required services not available");
		}

		if (!state) {
			state = {
				values: {},
				data: {},
				text: "",
			};
		}

		state.values = {
			...state.values,
			message: message.content.text,
			senderId: message.entityId,
			senderName: state.values?.senderName || "User",
			currentDateTime: new Date().toISOString(),
		};

		const prompt = composePromptFromState({
			state,
			template: scheduleFollowUpTemplate,
		});

		const response = await runtime.useModel(ModelType.TEXT_SMALL, {
			prompt,
			stopSequences: [],
		});

		const parsedResponse =
			parseKeyValueXml<ParsedScheduleFollowUpResponse>(response) ??
			(parseJSONObjectFromText(response) as ParsedScheduleFollowUpResponse | null) ??
			extractScheduleFollowUpResponseFromText(response);
		const contactName = parsedResponse?.contactName?.trim();
		if (!parsedResponse || (!contactName && !parsedResponse.entityId)) {
			logger.warn(
				"[ScheduleFollowUp] Failed to parse follow-up information from response",
			);
			throw new Error("Could not extract follow-up information");
		}

		let entityId = parsedResponse.entityId
			? asUUID(parsedResponse.entityId)
			: null;

		if (!entityId && contactName) {
			const contacts = await relationshipsService.searchContacts({
				searchTerm: contactName,
			});
			if (contacts.length > 0) {
				entityId = contacts[0]?.entityId ?? null;
			} else {
				const entity = await findEntityByName(runtime, message, state);
				if (entity?.id) {
					entityId = entity.id;
				} else {
					throw new Error(
						`Contact "${contactName}" not found in relationships`,
					);
				}
			}
		}

		if (!entityId) {
			throw new Error("Could not determine contact to follow up with");
		}

		const contact = await relationshipsService.getContact(entityId);
		if (!contact) {
			throw new Error(
				"Contact not found in relationships. Please add them first.",
			);
		}

		const scheduledAt = new Date(parsedResponse.scheduledAt || "");
		if (Number.isNaN(scheduledAt.getTime())) {
			throw new Error("Invalid follow-up date/time");
		}

		const task = await followUpService.scheduleFollowUp(
			entityId,
			scheduledAt,
			parsedResponse.reason || "Follow-up",
			(parsedResponse.priority as "high" | "medium" | "low") || "medium",
			parsedResponse.message,
		);

		const resolvedContactName = contactName || "contact";
		logger.info(
			`[ScheduleFollowUp] Scheduled follow-up for ${resolvedContactName} at ${scheduledAt.toISOString()}`,
		);

		const responseText = `I've scheduled a follow-up with ${resolvedContactName} for ${scheduledAt.toLocaleString()}. ${
			parsedResponse.reason ? `Reason: ${parsedResponse.reason}` : ""
		}`;

		if (callback) {
			await callback({
				text: responseText,
				action: "SCHEDULE_FOLLOW_UP",
				metadata: {
					contactId: entityId,
					contactName: resolvedContactName,
					scheduledAt: scheduledAt.toISOString(),
					taskId: task.id,
					success: true,
				},
			});
		}

		return {
			success: true,
			values: {
				contactId: entityId,
				taskId: task.id ?? "",
			},
			data: {
				contactId: entityId,
				contactName: resolvedContactName,
				scheduledAt: scheduledAt.toISOString(),
				taskId: task.id ?? "",
				reason: parsedResponse.reason ?? "",
				priority: parsedResponse.priority ?? "medium",
			},
			text: responseText,
		};
	},
};
